admin/tcl7.3/doc/Hash.3
'\" @OSF_COPYRIGHT@
'\" COPYRIGHT NOTICE
'\" Copyright (c) 1990, 1991, 1992, 1993, 1994, 1996 Open Software Foundation, Inc.
'\" ALL RIGHTS RESERVED (DCE).  See the file named COPYRIGHT.DCE for
'\" the full copyright text.
'\" 
'\" 
'\" HISTORY
'\" $Log: Hash.3,v $
'\" Revision 1.1.6.2  1996/03/09  23:07:41  marty
'\" 	Update OSF copyright year
'\" 	[1996/03/09  22:26:18  marty]
'\"
'\" Revision 1.1.6.1  1995/12/08  21:29:53  root
'\" 	Submit OSF/DCE 1.2.1
'\" 	[1995/12/08  18:01:32  root]
'\" 
'\" Revision 1.1.4.2  1994/06/10  20:56:02  devsrc
'\" 	cr10872- fix copyright
'\" 	[1994/06/10  16:21:59  devsrc]
'\" 
'\" Revision 1.1.4.1  1994/06/08  15:44:21  rousseau
'\" 	Change comment leaders and headers.
'\" 	[1994/06/08  15:41:14  rousseau]
'\" 
'\" $EndLog$
'\"
'\" Copyright (c) 1989-1993 The Regents of the University of California.
'\" All rights reserved.
'\"
'\" Permission is hereby granted, without written agreement and without
'\" license or royalty fees, to use, copy, modify, and distribute this
'\" documentation for any purpose, provided that the above copyright
'\" notice and the following two paragraphs appear in all copies.
'\"
'\" IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
'\" FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
'\" ARISING OUT OF THE USE OF THIS DOCUMENTATION, EVEN IF THE UNIVERSITY OF
'\" CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
'\"
'\" THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
'\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
'\" AND FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
'\" ON AN "AS IS" BASIS, AND THE UNIVERSITY OF CALIFORNIA HAS NO OBLIGATION TO
'\" PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
'\" 
'\" $Header: /u0/rcs_trees/dce/rcs/admin/tcl7.3/doc/Hash.3,v 1.1.6.2 1996/03/09 23:07:41 marty Exp $ SPRITE (Berkeley)
'\" 
.so man.macros
.HS Tcl_Hash tclc
.BS
.SH NAME
.na
Tcl_InitHashTable, Tcl_DeleteHashTable, Tcl_CreateHashEntry, Tcl_DeleteHashEntry, Tcl_FindHashEntry, Tcl_GetHashValue, Tcl_SetHashValue, Tcl_GetHashKey, Tcl_FirstHashEntry, Tcl_NextHashEntry, Tcl_HashStats \- procedures to manage hash tables
.SH SYNOPSIS
.nf
\fB#include <tcl.h>\fR
.sp
\fBTcl_InitHashTable\fR(\fItablePtr, keyType\fR)
.sp
\fBTcl_DeleteHashTable\fR(\fItablePtr\fR)
.sp
Tcl_HashEntry *
\fBTcl_CreateHashEntry\fR(\fItablePtr, key, newPtr\fR)
.sp
\fBTcl_DeleteHashEntry\fR(\fIentryPtr\fR)
.sp
Tcl_HashEntry *
\fBTcl_FindHashEntry\fR(\fItablePtr, key\fR)
.sp
ClientData
\fBTcl_GetHashValue\fR(\fIentryPtr\fR)
.sp
\fBTcl_SetHashValue\fR(\fIentryPtr, value\fR)
.sp
char *
\fBTcl_GetHashKey\fR(\fItablePtr, entryPtr\fR)
.sp
Tcl_HashEntry *
\fBTcl_FirstHashEntry\fR(\fItablePtr, searchPtr\fR)
.sp
Tcl_HashEntry *
\fBTcl_NextHashEntry\fR(\fIsearchPtr\fR)
.sp
char *
\fBTcl_HashStats\fR(\fItablePtr\fR)
.SH ARGUMENTS
.AS Tcl_HashSearch *searchPtr
.AP Tcl_HashTable *tablePtr in
Address of hash table structure (for all procedures but
\fBTcl_InitHashTable\fR, this must have been initialized by
previous call to \fBTcl_InitHashTable\fR).
.AP int keyType in
Kind of keys to use for new hash table.  Must be either
TCL_STRING_KEYS, TCL_ONE_WORD_KEYS, or an integer value
greater than 1.
.AP char *key in
Key to use for probe into table.  Exact form depends on
\fIkeyType\fR used to create table.
.AP int *newPtr out
The word at \fI*newPtr\fR is set to 1 if a new entry was created
and 0 if there was already an entry for \fIkey\fR.
.AP Tcl_HashEntry *entryPtr in
Pointer to hash table entry.
.AP ClientData value in
New value to assign to hash table entry.  Need not have type
ClientData, but must fit in same space as ClientData.
.AP Tcl_HashSearch *searchPtr in
Pointer to record to use to keep track of progress in enumerating
all the entries in a hash table.
.BE

.SH DESCRIPTION
.PP
A hash table consists of zero or more entries, each consisting of
a key and a value.
Given the key for an entry, the hashing routines can very quickly
locate the entry, and hence its value.
There may be at most one entry in a hash table with a
particular key, but many entries may have the same value.
Keys can take one of three forms:  strings,
one-word values, or integer arrays.
All of the keys in a given table have the same form, which is
specified when the table is initialized.
.PP
The value of a hash table entry can be anything that fits in
the same space as a ``char *'' pointer.
Values for hash table entries are managed entirely by clients,
not by the hash module itself.
Typically each entry's value is a pointer to a data structure
managed by client code.
.PP
Hash tables grow gracefully as the number of entries increases,
so that there are always less than three entries per hash bucket,
on average.
This allows for fast lookups regardless of the number of entries
in a table.
.PP
\fBTcl_InitHashTable\fR initializes a structure that describes
a new hash table.
The space for the structure is provided by the caller, not by
the hash module.
The value of \fIkeyType\fR indicates what kinds of keys will
be used for all entries in the table.  \fIKeyType\fR must have
one of the following values:
.IP \fBTCL_STRING_KEYS\fR 25
Keys are null-terminated ASCII strings.
They are passed to hashing routines using the address of the
first character of the string.
.IP \fBTCL_ONE_WORD_KEYS\fR 25
Keys are single-word values;  they are passed to hashing routines
and stored in hash table entries as ``char *'' values.
The pointer value is the key;  it need not (and usually doesn't)
actually point to a string.
.IP \fIother\fR 25
If \fIkeyType\fR is not TCL_STRING_KEYS or TCL_ONE_WORD_KEYS,
then it must be an integer value greater than 1.
In this case the keys will be arrays of ``int'' values, where
\fIkeyType\fR gives the number of ints in each key.
This allows structures to be used as keys.
All keys must have the same size.
Array keys are passed into hashing functions using the address
of the first int in the array.
.PP
\fBTcl_DeleteHashTable\fR deletes all of the entries in a hash
table and frees up the memory associated with the table's
bucket array and entries.
It does not free the actual table structure (pointed to
by \fItablePtr\fR), since that memory is assumed to be managed
by the client.
\fBTcl_DeleteHashTable\fR also does not free or otherwise
manipulate the values of the hash table entries.
If the entry values point to dynamically-allocated memory, then
it is the client's responsibility to free these structures
before deleting the table.
.PP
\fBTcl_CreateHashEntry\fR locates the entry corresponding to a
particular key, creating a new entry in the table if there
wasn't already one with the given key.
If an entry already existed with the given key then \fI*newPtr\fR
is set to zero.
If a new entry was created, then \fI*newPtr\fR is set to a non-zero
value and the value of the new entry will be set to zero.
The return value from \fBTcl_CreateHashEntry\fR is a pointer to
the entry, which may be used to retrieve and modify the entry's
value or to delete the entry from the table.
.PP
\fBTcl_DeleteHashEntry\fR will remove an existing entry from a
table.
The memory associated with the entry itself will be freed, but
the client is responsible for any cleanup associated with the
entry's value, such as freeing a structure that it points to.
.PP
\fBTcl_FindHashEntry\fR is similar to \fBTcl_CreateHashEntry\fR
except that it doesn't create a new entry if the key doesn't exist;
instead, it returns NULL as result.
.PP
\fBTcl_GetHashValue\fR and \fBTcl_SetHashValue\fR are used to
read and write an entry's value, respectively.
Values are stored and retrieved as type ``ClientData'', which is
large enough to hold a pointer value.  On almost all machines this is
large enough to hold an integer value too.
.PP
\fBTcl_GetHashKey\fR returns the key for a given hash table entry,
either as a pointer to a string, a one-word (``char *'') key, or
as a pointer to the first word of an array of integers, depending
on the \fIkeyType\fR used to create a hash table.
In all cases \fBTcl_GetHashKey\fR returns a result with type
``char *''.
When the key is a string or array, the result of \fBTcl_GetHashKey\fR
points to information in the table entry;  this information will
remain valid until the entry is deleted or its table is deleted.
.PP
\fBTcl_FirstHashEntry\fR and \fBTcl_NextHashEntry\fR may be used
to scan all of the entries in a hash table.
A structure of type ``Tcl_HashSearch'', provided by the client,
is used to keep track of progress through the table.
\fBTcl_FirstHashEntry\fR initializes the search record and
returns the first entry in the table (or NULL if the table is
empty).
Each susequent call to \fBTcl_NextHashEntry\fR returns the
next entry in the table or
NULL if the end of the table has been reached.
A call to \fBTcl_FirstHashEntry\fR followed by calls to
\fBTcl_NextHashEntry\fR will return each of the entries in
the table exactly once, in an arbitrary order.
It is unadvisable to modify the structure of the table, e.g.
by creating or deleting entries, while the search is in
progress.
.PP
\fBTcl_HashStats\fR returns a dynamically-allocated string with
overall information about a hash table, such as the number of
entries it contains, the number of buckets in its hash array,
and the utilization of the buckets.
It is the caller's responsibility to free the result string
by passing it to \fBfree\fR.
.PP
The header file \fBtcl.h\fR defines the actual data structures
used to implement hash tables.
This is necessary so that clients can allocate Tcl_HashTable
structures and so that macros can be used to read and write
the values of entries.
However, users of the hashing routines should never refer directly
to any of the fields of any of the hash-related data structures;
use the procedures and macros defined here.

.SH KEYWORDS
hash table, key, lookup, search, value


file/util/hash.c
/*
 * @OSF_COPYRIGHT@
 * COPYRIGHT NOTICE
 * Copyright (c) 1990, 1991, 1992, 1993, 1994 Open Software Foundation, Inc.
 * ALL RIGHTS RESERVED (DCE).  See the file named COPYRIGHT.DCE for
 * the full copyright text.
 */
/*
 * HISTORY
 * $Log: hash.c,v $
 * Revision 1.1.323.1  1996/10/02  21:13:11  damon
 * 	Newest DFS from Transarc
 * 	[1996/10/01  18:51:01  damon]
 *
 * Revision 1.1.318.2  1994/06/09  14:24:49  annie
 * 	fixed copyright in src/file
 * 	[1994/06/09  13:36:54  annie]
 * 
 * Revision 1.1.318.1  1994/02/04  20:35:34  devsrc
 * 	Merged from 1.0.3a to 1.1
 * 	[1994/02/04  15:20:18  devsrc]
 * 
 * Revision 1.1.316.1  1993/12/07  17:37:46  jaffe
 * 	1.0.3a update from Transarc
 * 	[1993/12/03  18:08:32  jaffe]
 * 
 * $EndLog$
*/
/* This is a general purpose hash table package.  A hash table typically
 * declared as a static variable by the user of the package.  It must be
 * initialized and a key comparison method must be specified.  Entries are
 * structures which are passed by pointer (coerced to (dfsh_hashEntry_t)).
 * Each entry is required to include a thread pointer (sizeof
 * (dfsh_hashEntry_t)) at a fixed offset.  A function that takes a pointer to
 * an entry an returned a hash value (unsigned long) is also required. */

/* Copyright (C) 1994, 1989 Transarc Corporation - All rights reserved */

/* For an example of the package's usage see ktc/ktc.c */

/* Moved from afs/ktc package to DEcorum. 900321 -ota */
/* Moved from de/episode/anode to de/episode/tools.  910515 -wam */
/* Copied from src/file/episode/tools to src/file/util 930330 -jdp */

#include <dcedfs/param.h>
#include <dcedfs/osi.h>
#include <dcedfs/stds.h>

#include "hash.h"

RCSID ("$Header: /u0/rcs_trees/dce/rcs/file/util/hash.c,v 1.1.323.1 1996/10/02 21:13:11 damon Exp $")

#define UNSPECIFIED 0
#define FIXED 1
#define VARYING 2
#define STRING 3
#define FUNCTION 4

#if (DFSH_DEBUG>5)
#define osi_DebugMsg(a) printf a
#define HT_DEBUG(name,ht) printf ("HT in %s @%x %x %d %d\n", name, ht, (ht)->hashFunction, (ht)->entries, (ht)->length)
#else
#define osi_DebugMsg(a)
#define HT_DEBUG(in,ht)
#endif

/* Very cheap locking.  0 is unlocked, positive is number of lockers, -1 is
 * write locked.  Note that panic is a macro that turns into abort, which is a
 * void returning function.  This causes problems with the ?: operator on some
 * compilers. */

#if defined(DFSH_HASH_LOCK)
#define InitLock(lockP) (*(lockP) = 0)
#define HeldLockP(lockP) (*(lockP) != 0)

#define ReadLock(lockP) \
    ( (*(lockP) < 0) ? (panic ("write lock held"),0) : (*(lockP))++ )

#define ReadUnlock(lockP) \
    ( (*(lockP) <= 0) ? (panic ("no read lock held"),0) : (*(lockP))-- )

#define WriteLock(lockP) \
    ( (*(lockP) != 0) ? (panic ("lock held"),0) : (*(lockP))-- )

#define WriteUnlock(lockP) \
    ( (*(lockP) != -1) ? (panic ("no write lock held"),0) : (*(lockP))++ )

/* convert a write lock to a read lock */
#define DowngradeLock(lockP) \
    ( (*(lockP) != -1) ? (panic ("no write lock held"),0) : ((*(lockP)) = 1) )
#else
#define InitLock(lockP)
#define HeldLockP(lockP)
#define ReadLock(lockP)
#define ReadUnlock(lockP)
#define WriteLock(lockP)
#define WriteUnlock(lockP)
#define DowngradeLock(lockP)
#endif /* DFSH_HASH_LOCK */

/* dfsh_HashInit - initializes a hash table.  It takes a pointer to the hash
 * table, a function that returns an entry's hash value, and the offset (in
 * bytes) in the entry of the hash thread pointer. */

/* EXPORT */
int dfsh_HashInit (
  IN struct dfsh_hashTable *ht,
  IN unsigned long (*hf)(),
  IN int to)
{
    HT_DEBUG("dfsh_HashInit in",ht);
    if (!ht || !hf) return -1;
    bzero ((char *)ht, sizeof(*ht));
    ht->hashFunction = hf;
    ht->threadOffset = to;
    ht->keyType = UNSPECIFIED;
    InitLock (&ht->lock);
    HT_DEBUG("dfsh_HashInit out",ht);
    return 0;
}

/* dfsh_HashInitKey* - this is a family of functions that initializes the entry
 * comparison function for a hash table.  Each takes a pointer to the hash
 * table and one or more arguments which describe the comparison to make.  This
 * allows the lookup function to find the matching entry. */

/* dfsh_HashInitKeyFixed - the key is described by offset in the entry and
 * length in bytes.  This is best used if there are one or more fixed size
 * fields, such as pointers or integers, to be checked. */

/* EXPORT */
long dfsh_HashInitKeyFixed (
  struct dfsh_hashTable *ht,
  int off,
  int len)
{
    ht->keyType = FIXED;
    ht->cmpData.fixed.off = off;
    ht->cmpData.fixed.len = len;
    return 0;
}

/* dfsh_HashInitKeyVarying - the key is described as an offset to an int which
 * provides the length of an immediately following sequence of bytes. This is
 * used for counted byte strings which are not null terminated.  The comparison
 * string is bcmp. */

/* EXPORT */
long dfsh_HashInitKeyVarying (
  struct dfsh_hashTable *ht,
  int off)
{
    ht->keyType = VARYING;
    ht->cmpData.varying = off;
    return 0;
}

/* dfsh_HashInitKeyString - the key is a string whose location is given by its
 * offset in the entry.  This is for regular null terminated C strings.  The
 * comparison function used is strcmp. */

/* EXPORT */
long dfsh_HashInitKeyString (
  struct dfsh_hashTable *ht,
  int off)
{
    ht->keyType = STRING;
    ht->cmpData.string = off;
    return 0;
}

/* dfsh_HashInitKeyFunction - the key cannot be specified by any of the above
 * methods so this provides a boolean function which takes two entries and
 * compares them for equality. */

/* EXPORT */
long dfsh_HashInitKeyFunction (
  struct dfsh_hashTable *ht,
  int (*fun)())
{
    ht->keyType = FUNCTION;
    ht->cmpData.fun = fun;
    return 0;
}

#define INITIAL_TABLE_SIZE 17

static int primes[] = {3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47,
	53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 0};

static int GetBiggerTable (int old)
{
    int  new;
    int *pl;

    new = old * 3;			/* start about 1/3 full */
    for (new |= 1; ; new += 2) {	/* force it odd, look for primes */
	for (pl = primes; *pl; pl++) {
	    int rem = new / (*pl);
	    if (rem * (*pl) == new) goto not_prime;
	    if (rem <= (*pl)) break;
	}
	return new;
      not_prime:
	;
    }
}

static int GetSmallerTable (int len, int ent)
{
    int  new;
    int *pl;

    if ((ent == 0) && (len < 2*INITIAL_TABLE_SIZE)) return len;
    new = ent * 2;			/* start about half full */
    /* force it odd, look for primes */
    for (new |= 1; new > INITIAL_TABLE_SIZE; new -= 2) {
	for (pl = primes; *pl; pl++) {
	    int rem = new / (*pl);
	    if (rem * (*pl) == new) goto not_prime;
	    if (rem <= (*pl)) break;
	}
	return new;
      not_prime:
	;
    }
    return INITIAL_TABLE_SIZE;
}

static void
Rehash(struct dfsh_hashTable *ht, dfsh_hashEntry_t *otable, int olength)
{
    HT_DEBUG("rehash",ht);
    if (ht->length == olength) {
	return;
    } else if (otable == 0) {
	afsl_Assert(ht->length == 0);
	return;
    } else if (ht->length == 0) {
	afsl_Assert(ht->entries == 0);
	ht->emptyCnt++;			/* hash table empty; being discarded */
	osi_DebugMsg (("freeing %d word hash table\n", olength));
	ht->table = 0;
    } else {
	int i;

	ht->rehashCnt++;
	if (ht->length < olength) {
	    osi_DebugMsg (("Shrinking %d word hash table to %d words\n", olength, ht->length));
	} else {
	    osi_DebugMsg (("Growing %d word hash table to %d words\n", olength, ht->length));
	}
	ht->table = (dfsh_hashEntry_t *)
	    osi_Alloc_r(ht->length * sizeof (dfsh_hashEntry_t));
	for (i = 0; i < ht->length; i++)
	     ht->table[i] = 0;
	ht->entries = 0;
	for (i = 0; i < olength; i++) {
	    dfsh_hashEntry_t e;
	    dfsh_hashEntry_t next;

	    for (e = otable[i]; e; e = next) {
		next = *(dfsh_hashEntry_t *)(e + ht->threadOffset);
		dfsh_HashIn_r(ht, e);
	    }
	}
	ht->hashInCnt -= olength;
    }
    osi_Free_r((char *)otable, olength*sizeof(dfsh_hashEntry_t));
}

/*
 * dfsh_HashIn -- adds an entry to a hash table.  It takes a pointer to the
 *     hash table and a pointer to the entry.  A value of -1 is returned if the
 *     entry is already in the hash table.
 *
 * dfsh_HashIn_r is the preemption-safe version.
 */
int
dfsh_HashIn(struct dfsh_hashTable *ht, opaque ahe)
{
    int code;
    osi_MakePreemptionRight();
    code = dfsh_HashIn_r(ht, ahe);
    osi_UnmakePreemptionRight();
    return code;
}

int
dfsh_HashIn_r(struct dfsh_hashTable *ht, opaque ahe)
{
    unsigned long hash;
    int hi;
    dfsh_hashEntry_t he = ahe;
    dfsh_hashEntry_t e;
    dfsh_hashEntry_t *next;

#if (DFSH_DEBUG>10)
    HT_DEBUG("HashIn",ht);
#endif
    ht->hashInCnt++;
    if (ht->length == 0) {
	int i;
	ht->length = INITIAL_TABLE_SIZE;
	ht->table = (dfsh_hashEntry_t *)
	    osi_Alloc_r(ht->length * sizeof (dfsh_hashEntry_t));
	for (i = 0; i < ht->length; i++)
	    ht->table[i] = 0;
    } else if (ht->entries > 2 * ht->length) {
	dfsh_hashEntry_t *otable = ht->table;
	int olength = ht->length;

	ht->length = GetBiggerTable (ht->entries);
	Rehash(ht, otable, olength);
    }

    hash = (*ht->hashFunction)(he);
    hi = hash % ht->length;

#if (DFSH_DEBUG>5)
    /* check for entry already on list, to avoid loops */
    for (e = ht->table[hi]; e;
	 e = *(dfsh_hashEntry_t *)(e + ht->threadOffset))
	if (e == he) return -1;
#endif					/* DFSH_DEBUG */

    next = (dfsh_hashEntry_t *)(he + ht->threadOffset);
    *next = ht->table[hi];
    ht->table[hi] = he;
    ht->entries++;
    return 0;
}

/*
 * dfsh_HashOut -- removes an entry from a hash table.  It takes a pointer to
 *     the hash table and a pointer to the entry.  A value of -1 is returned if
 *     the entry was not in the hash table.
 *
 * dfsh_HashOut_r is the preemption-safe version.
 */
int
dfsh_HashOut(struct dfsh_hashTable *ht, opaque ahe)
{
    int code;
    osi_MakePreemptionRight();
    code = dfsh_HashOut_r(ht, ahe);
    osi_UnmakePreemptionRight();
    return code;
}

int
dfsh_HashOut_r(struct dfsh_hashTable *ht, opaque ahe)
{
    u_long hash;
    int hi, result = 0;
    dfsh_hashEntry_t he = ahe;
    dfsh_hashEntry_t e;
    dfsh_hashEntry_t *next;

    HT_DEBUG("dfsh_HashOut",ht);
    afsl_Assert(ht != NULL && he != NULL);
    WriteLock (&ht->lock);
    afsl_Assert(ht->table != NULL);

    ht->hashOutCnt++;

    hash = (*ht->hashFunction)(he);
    hi = hash % ht->length;

    /* Find entry on list */
    next = (dfsh_hashEntry_t *)(he + ht->threadOffset);
    if (he == ht->table[hi])
	ht->table[hi] = *next;
    else {
	dfsh_hashEntry_t last_e = 0;
	for (e = ht->table[hi]; e != he && e != NULL;
	    e = *(dfsh_hashEntry_t *)(e + ht->threadOffset)) {
	    last_e = e;
	}
	if (e == NULL) {
	    result = -1;
	    goto done;
	}
	*(dfsh_hashEntry_t *)(last_e + ht->threadOffset) = *next;
    }

    ht->entries--;
    if (ht->entries * 8 < ht->length) {
	dfsh_hashEntry_t *otable = ht->table;
	int olength = ht->length;

	ht->length = GetSmallerTable (ht->length, ht->entries);
	Rehash(ht, otable, olength);
    }
done:
    WriteUnlock (&ht->lock);
    return result;
}

/* dfsh_HashLookup - locates an entry in a hash table and returns a pointer to
 * it.  It takes a hash table pointer and a pointer to a dummy entry which
 * serves as a comparison key.  The pointer is returned to the first entry that
 * matches the key entry.  If no entry is found a zero pointer is returned. */

/* EXPORT */
opaque dfsh_HashLookup (
  struct dfsh_hashTable *ht,
  opaque akey)
{
    unsigned long hash;
    int hi;
    dfsh_hashEntry_t key = akey;
    dfsh_hashEntry_t e;
    int off;				/* offset */
    int len;				/* length of compare */

    HT_DEBUG("dfsh_HashLookup",ht);
    if (!ht || !key) panic ("bad call to HashLookup");
    ReadLock (&ht->lock);
    e = 0;
    if (ht->length == 0) goto done;

    hash = (*ht->hashFunction)(key);
    hi = hash % ht->length;

    switch (ht->keyType) {
      case FIXED:
	off = ht->cmpData.fixed.off;
	len = ht->cmpData.fixed.len;
	for (e = ht->table[hi]; e; e = *(dfsh_hashEntry_t *)(e+ht->threadOffset))
	    if (bcmp (key+off, e+off, len) == 0) goto done;
	break;
      case VARYING:
	off = ht->cmpData.varying;
	len = *(int *)(key+off);
	off += sizeof(int);
	for (e = ht->table[hi]; e; e = *(dfsh_hashEntry_t *)(e+ht->threadOffset))
	    if ((len == *(int *)(e + off-sizeof(int))) &&
		(bcmp (key+off, e+off, len) == 0)) goto done;
	break;
      case STRING:
	off = ht->cmpData.string;
	for (e = ht->table[hi]; e; e = *(dfsh_hashEntry_t *)(e+ht->threadOffset))
	    if (strcmp (key+off, e+off) == 0) goto done;
	break;
      case FUNCTION:
	for (e = ht->table[hi]; e; e = *(dfsh_hashEntry_t *)(e+ht->threadOffset))
	    if ((*ht->cmpData.fun) (key, e)) goto done;
	break;
      default: panic ("no key type");
    }
  done:
    ReadUnlock (&ht->lock);
    return e;
}

/* This procedure can be used to enumerate the contents of a hash table.  On
 * the first call he is zero and the first hash entry is returned.  On
 * subsequent calls the entry returned by the previous call is provided and the
 * next entry is returned.
 *
 * NOTE that the hash table must not be modified during this sequence as
 * modifications may result in the hash table being reorganized.  The can
 * result in duplicate and omitted entries. */

/* EXPORT */
opaque dfsh_HashNext (
  struct dfsh_hashTable *ht,
  opaque ahe)
{
    dfsh_hashEntry_t he = ahe;
    dfsh_hashEntry_t e = 0;
    int hi;
    dfsh_hashEntry_t next;

#if (DFSH_DEBUG>10)
    HT_DEBUG("dfsh_HashNext",ht);
#endif
    if (ht == 0) panic ("bad call to HashNext");
    ReadLock (&ht->lock);
    if (ht->length == 0) goto done;	/* hash table empty */
    if (he == 0) {
	hi = -1;			/* start at beginning of hash table */
	next = 0;
    } else {
	unsigned long hash = (*ht->hashFunction)(he);
	hi = hash % ht->length;
	next = *(dfsh_hashEntry_t *)(he + ht->threadOffset);
    }
    if (next == 0) {			/* find next non-empty bucket */
	for (hi++; hi < ht->length; hi++) {
	    e = ht->table[hi];
	    if (e) goto done;
	}
	e = 0;
	goto done;			/* no more entries */
    }
    /* Make sure next if really in this bucket, to prevent indirecting through
     * garbage parameter. */
    for (e = ht->table[hi]; e; e = *(dfsh_hashEntry_t *)(e + ht->threadOffset))
	if (e == next) goto done;
    /* input parameter was garbage, panic? */
  done:
    ReadUnlock (&ht->lock);
    return e;
}
